jcl/src/java.base/share/classes/openj9/internal/foreign/abi/InternalDowncallHandler.java

```diff
@@ -123,7 +125,18 @@ void clear() {
 		}
 	}
 
-	private final ThreadLocal<HeapArgInfo> heapArgInfo;
+	/* heapArgInfo holds a per-thread stack of HeapArgInfo structures.
+	 * Each entry tracks Java references to heap-allocated memory segments
+	 * Required for recursive downcalls.
+	 * Each downcall maintains its own set of references until completion.
+	 *
+	 * createHeapArgInfo controls when a new entry should be created.
+	 * If true, a new HeapArgInfo is pushed for the current downcall.
+	 * A single downcall may have multiple pointer arguments,
+	 * and all of them share the same entry.
+	 */
+	private static final ThreadLocal<Deque<HeapArgInfo>> heapArgInfo = ThreadLocal.withInitial(ArrayDeque::new);
```

Comment:
I think this can be simplified:

1) Always create a  `HeapArgInfo` in `InternalDowncallHandler(MethodType functionMethodType, FunctionDescriptor functionDescriptor)` and push it on to the local stack. And always pop it after the call irrespective of whether pointers are used or not. This removes the need for `createHeapArgInfo`.

2) In the constructor for `HeapArgInfo` do not allocate the `bases` and `offsets` upfront, instead check and allocate when `append()` is called.

3) All the `if (info == null) {` checks can be removed. However, the `if (info != null) {` still need to be kept.